qiskit_machine_learning/kernels/fidelity_quantum_kernel.py

```diff
@@ -0,0 +1,289 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Fidelity Quantum Kernel"""
+
+from __future__ import annotations
+
+import warnings
+from typing import List, Tuple
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.algorithms.state_fidelities import BaseStateFidelity, ComputeUncompute
+from qiskit.primitives import Sampler
+
+from qiskit_machine_learning.kernels.base_kernel import BaseKernel
+
+KernelIndices = List[Tuple[int, int]]
+
+
+class FidelityQuantumKernel(BaseKernel):
+    r"""
+    QuantumKernel
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    .. math::
+
+        K(x, y) = \langle f(x), f(y)\rangle.
+
+    Here K is the kernel function, x, y are n dimensional inputs. f is a map from n-dimension
+    to m-dimension space. :math:`\langle x, y \rangle` denotes the dot product.
+    Usually m is much larger than n.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints x and y and feature
+    map f, all of n dimension. This kernel matrix can then be used in classical machine learning
+    algorithms such as support vector classification, spectral clustering or ridge regression.
+
+    Here, the kernel function is defined as the overlap of two quantum states defined by a
+    parametrized quantum circuit (called feature map):
+
+    .. math::
+
+        K(x,y) = |\langle \phi(x) | \phi(y) \rangle|^2
+    """
+
+    def __init__(
+        self,
+        sampler: Sampler | None = None,
+        feature_map: QuantumCircuit | None = None,
+        fidelity: str | BaseStateFidelity = "zero_prob",
```

Comment:
Do you see doing other string values later? If other fidelities have options then strings are not really going to cut it. I have to wonder why not just have the user pass a fidelity instance and drop the sampler. If you want to default the fidelity to ComputeUncompute with the reference Sampler that seems fine. The duality of Sampler and Fidelity here just seems to complicate things for me.

Reply:
I think the idea was that this way allows users to use the kernel very similarly to how they used the old QuantumKernel. Instead of a quanutm instance, they now just provide a sampler and don't have to care about how fidelities are calculated. 

I agree that this complicates the code a lot and maybe should be discussed again.

Reply:
Maybe convenience/factory method(s) could be provided. Have one that takes a Sampler and builds out the fidelity with it and returns an instance. While a user could do that, if its something they often want to do a separate factory method could handle that rather than having a more complicated constructor that incorporates that too.

Reply:
I guess it is fine to have just `fidelity` as a parameter. The code is simpler, but less flexible from the user perspective.